presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java

```diff
@@ -92,6 +92,7 @@ public void noMoreOperators()
     private final PlanNodeId sourceId;
     private final ExchangeClient exchangeClient;
     private final PagesSerde serde;
+    private ListenableFuture<?> isBlocked = NOT_BLOCKED;
```

Comment:
All accesses of this field (and operators in general) is done while holding the driver lock in a single threaded manner, so unsafe publication shouldn't be a concern. Even if that weren't true, the only risk would be somehow seeing `null` in this field because all of the `ListenableFuture<?>` instances returned here are amenable to being called multiple times- just at the cost of synchronizing into a critical section.

Reply:
I don't think there is a guarantee that the ListenableFuture implementation is constructed in such a way that after the constructor completes the fields are visible? I did look at all the callers and yeah it looks like it is all from the driver loop I just didn't have as high a confidence as you that I wasn't missing something.

LGTM